components/cosmetic_filters/renderer/cosmetic_filters_js_handler.cc

```diff
@@ -116,6 +116,51 @@ const char kHideSelectorsInjectScript[] =
           };
         })();)";
 
+const char kRemovalsInjectScript[] =
+    R"((function() {
+          const selectorsToRemove = %s;
+          const classesToRemoveBySelector = %s;
+          const attributesToRemoveBySelector = %s;
+          const onMutation = (mutationList) => {
+            if (mutationList.find(
+                mutation => (
+                  mutation.addedNodes.length !== 0 && mutation.addedNodes.values().find(node => node.nodeType == 1)
+                )
+              )
+            ) {
+              execute();
+            }
+          };
+          const execute = () => {
+              selectorsToRemove.forEach(
```

Comment:
I wonder why we call `querySelectorAll` instead using the list of nodes provided by the mutation observer

Reply:
CSS selectors can refer to parent / child / adjacent nodes... do we need to care about that?

Reply:
There are two approaches:
1. the event based, when we check only modified nodes. Works perfectly when number of DOM changes is limited.
2. Polling. When check the whole DOM from time to time.

In average, 1 better in terms of perf, but it has some corner cases.
If we use 2, we should be careful with CPU efficiency: a) the number of execute() calls a) the number of `querySelectorAll` calls inside `execute()`.

For a) I assume there is a throttling policy in MutationObserver, but it should be verified if we want to relay on it.

Reply:
If you have a list of mutated elements, how do you check if they match a selector?

Reply:
using [element.matches()](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches)?

Reply:
> using [element.matches()](https://developer.mozilla.org/en-US/docs/Web/API/Element/matches)?

Thank you!